### DISCUSSION TITLE
How to Get Git Out of my User Folder

### DISCUSSION THREAD

**Discussion by AdreKiseque** [git-for-windows/git#5183] [Q&A] [answered]
So, currently, Git likes to use the main user folder (C:\Users\Username) as a dumping ground for a variety of its files... I can't name all of the ones I've seen but right now I have specimens such as `.bash_history`, `.lesshst` and of course `.gitconfig`. I also recall getting a "sources" folder in a previous environment and I believe I've had at least one more folder in here as well. Needless to say, I don't appreciate my computer's equivelant to a foyer being used as a gratuitous AppData directory, and would much rather have Git drop its configuration files in... well most ideally the actual AppData directory. I know about setting the `GIT_CONFIG_GLOBAL` envar for `.gitconfig` (though I haven't been able to get it actually working yet due to some permissions issues), but it doesn't address the various other things that are dropped in here. Does anyone know if it's possible to fully redirect Git out of here and into somewhere else?

I'd actually like to open an issue for this topic, to suggest a new default location since putting things straight into the User folder without at least another container is really just absurd and messy, but I'm not sure what the proper avenue would be. As far as I can tell, the issues tracker on this repository is only for actual bugs, and not feature suggestions? If someone could direct me, it would be appreciated.

**Comment by dscho** [comment, marked as answer]
The easiest method would be to define your `HOME` environment variable to point somewhere else.

> I'd actually like to open an issue for this topic, to suggest a new default location

This would break everybody else's existing setups, and is therefore a complete non-starter. I'm not sure that you even considered the consequences on other users?

**Comment by AdreKiseque** [comment]
> The easiest method would be to define your `HOME` environment variable to point somewhere else.

Except the `HOME` envar affects things all over my computer, which I don't necessarily want changed as well.
> This would break everybody else's existing setups, and is therefore a complete non-starter. I'm not sure that you even considered the consequences on other users?

This isn't true at all. A new default setting at installation would have no impact on existing setups, just like how changing the default branch name doesn't affect existing repositories. What it would do is make new installations a lot more organized, and fascilitate a shift over the long term to a more manageable setup.
In parallel to the default repo name, actually, this should ideally be something users can configure during setup, giving them the option to choose a location off the bat as well as to modify the value afterwards without having to mess with global environment variables. It seems reasonable, doesn't it?

**Reply by dscho** [reply]
Did you mean to reply to https://github.com/git-for-windows/git/discussions/5183#discussioncomment-10805499? I ask because it looks as if you tried to provide an alternative answer, but this is not like an alternative answer at all.

As to changing the defaults: If there was no existing documentation out there, there would be merit to that argument. But there is. A lot.

**Reply by AdreKiseque** [reply]
I did mean to reply, yes. Sorry for the hostility, I got a bit defensive.

However, I disagree the current state of things is reasonable. For starters, the default configuration is what I can only describe as sloppy. Programs storing their files directly in the User folder instead of AppData is one thing, but Git drops them directly in the directory as loose files, just making a mess. The fact it can only be easily changed through a generic `HOME` variable, which may be used by other apps as well, is also a problem.

(As a note, while I do recall finding mention of being able to change the `HOME` variable, it doesn't seem to be on the [doc page](https://git-scm.com/docs/git-config#ENVIRONMENT) itself. Maybe I'm looking in the wrong place?)

A possible solution I might propose, maintaining full compatibility whilst also addressing all of the issues: Create a new variable specifically for Git such as `GIT_HOME`, serving the current use of `HOME` for Git and defaulting to it if not set. This would allow users better control of Git's installation without breaking existing setups. Further, this variable could be set automatically during initial installation (perhaps even with a user prompt for customization) to a more organized default location, such as `User\AppData\Roaming\Git`. Again, this would make new installations more organized and clean without having any impact on existing setups, since it would only happen one newly installed instances.

Thank you for taking the time to respond to me.

**Reply by dscho** [reply]
It might be seen as quite daring to call other people's work "sloppy" when having _just_ been called out for not replying to the correct message.

In any case, I still do not agree that the added complication is worth the benefit. Those proposed work-arounds are _prone_ to break in (not actually) entertaining ways. It would appear that the ship has sailed to change the location of the `.gitconfig` file, for example, which many a user track via so-called `dotfiles` repositories, for one. And that's just one example. It is easy to over-index on one's own usage scenarios, but Git for Windows is used by _millions_ of users. Whatever you think is "unlikely in practice" is actually happening, and not only a few dozen times.

So no, in the interest of keeping this software usable for users who have grown accustomed to it working for them in their scenarios, I will not change the default. As I mentioned earlier, this idea is a non-starter. You can disagree with me, that's totally legitimate. But it would take much more to convince me to break existing users' setups than a mere opinion.

**Reply by AdreKiseque** [reply]
That's fair, thank you for your time.
